app/domain/authentication/authn_k8s/k8s_resolver.rb

```diff
@@ -65,101 +55,116 @@ def pod_owner_refs
 
         # Validates that the +pod+ belongs to the resource object.
         #
-        # @exception ValidationError if the +pod+ does not belong to the resource object.
+        # @exception PodNameMismatchError,PodRelationMismatchError, PodMissingRelationError if the +pod+ does not
+        # belong to the resource object.
         def validate_pod
           raise "validate_pod is not implemented"
         end
       end
 
       # Tests whether the Pod's ReplicaSet belongs to the Deployment.
       class Deployment < Base
+        def replica_set_owner_refs(replica_set)
+          @replica_set_owner_refs ||= replica_set.metadata.ownerReferences
+        end
+
         def validate_pod
-          replica_set_ref = verify "Pod #{pod_name} does not belong to a ReplicaSet (or Deployment)" do
-            pod_owner_refs &&
-              pod_owner_refs.find{|ref| ref.kind == "ReplicaSet"}
+          replica_set_ref = pod_owner_refs.find{|ref| ref.kind == "ReplicaSet"}
+          unless pod_owner_refs && replica_set_ref
+            raise Err::PodMissingRelationError.new(pod_name, 'ReplicaSet')
           end
```

Comment:
this code is also problematic. You first call `pod_owner_refs.find{|ref| ref.kind == "ReplicaSet"}` and only then verify that `pod_owner_refs` is not nil.